src/tests/prepare.spec.js

```diff
@@ -113,6 +113,44 @@ describe('prepare', () => {
     t.assert(doAsyncSideEffect.calledThrice, 'Should be called 3 times');
   });
 
+  it('Should handle data deps properly in correct order', async () => {
+    const execOrder = [];
+
+    const innerFunc = () => execOrder.push('inner');
+    const outerFunc = () => execOrder.push('outer');
+
+    const outerPrepare = async () =>
+      new Promise(resolve => setTimeout(() => outerFunc() && resolve(), 0));
+    const innerPrepare = async () =>
+      new Promise(resolve => innerFunc() && resolve());
+
+    const Outer = prepared(outerPrepare)(
+      ({ text, children }) => (
+        <div>
+          {text} <div>{children ? children : null}</div>
+        </div>
+      ),
+      { hasSsrDataDeps: false },
+    );
+    const Inner = prepared(innerPrepare)(({ text, children }) => (
+      <div>
+        {text} <div>{children ? children : null}</div>
+      </div>
+    ));
+    await prepare(
+      <Outer text="foo">
+        <Inner text="foo">
+          <Inner text="foo" />
+          <Inner text="foo" />
+        </Inner>
+      </Outer>,
+    );
+    t.assert(
+      equal(execOrder, ['inner', 'inner', 'inner', 'outer']),
+      'outer should be resolved last',
```

Comment:
If the outer element `hasSsrDataDeps: false`, shouldn't it then be resolved first? Or is that not the point here? We are a little confused 😕  🙃  We kinda see how it works, but we're a little confused with the comment.

![image](https://user-images.githubusercontent.com/23152018/96016528-26208900-0e49-11eb-9dd1-bbcd1228b7cb.png)

The question is really, where does the speed come from :)

Reply:
Okey we kinda maybe get it now. Correct us if the following logic is wrong.

If the outer prepared function has no deps, then there is no need to await the result before traversing down to the inner elements. Then the outer prepared promise is appended to the end, and Promise.all will make sure that is awaited and resolved before the rendering the html. Thus avoiding an await at each level of the tree?

Reply:
I think I get this now. Correct me if I am wrong but the `preparedComponent` with `dataDeps = false` is actually running in sync with the `prepareFunc` of the closest child `preparedComponent`? (and etc. for all consecutive children with no deps?)

Reply:
Hahah, yeah, the naming scheme is quite bad. Should probably be something like `dontWaitForPrepareFuncToFinishBeforeTraversingFurtherButWaitAtTheEnd`. So yeah, that is right @LudvigHz. Eg. when fetching readme-as-function, we can keen traversing further and start fetching other stuff, since no fetch logic depends on the data from readme-as-a-function. But we have to wait for fetchLoggedInUser to succeed in before traversing because we need the user id for some other fetch logic eg.

So if you have a better name feel free to suggest. (thought about using `sync` as well, but the actions are async, so sync would probably make things harder to understand).


> The question is really, where does the speed come from :)

This is now how actions on lego-webapp runs on ssr with this patch:

```                              
2020-10-14T15:34:48.968Z User.LOGIN.SUCCESS                                                                                                                     
2020-10-14T15:34:48.978Z User.FETCH.BEGIN                                                                                                                       
2020-10-14T15:34:48.982Z Meta.FETCH.BEGIN                                                                                                                       
2020-10-14T15:34:48.987Z Meeting.FETCH.BEGIN                                                                                                                    
2020-10-14T15:34:49.023Z Quote.FETCH_RANDOM.BEGIN                                                                                                               
2020-10-14T15:34:49.024Z Readme.FETCH.BEGIN                                                                                                                     
2020-10-14T15:34:49.030Z Frontpage.FETCH.BEGIN                                                                                                                  
2020-10-14T15:34:49.398Z Meta.FETCH.SUCCESS                                                                                                                     
2020-10-14T15:34:49.415Z Readme.FETCH.SUCCESS                                                                                                                   
2020-10-14T15:34:49.434Z Meeting.FETCH.SUCCESS                                                                                                                  
2020-10-14T15:34:49.552Z User.FETCH.SUCCESS                                                                                                                     
2020-10-14T15:34:49.594Z Quote.FETCH_RANDOM.SUCCESS                                                                                                             
2020-10-14T15:34:49.857Z Frontpage.FETCH.SUCCESS
15:34:49.953Z  INFO lego-webapp: request (method=GET, url=/, status=200, content-length=209329, response-time=989.647
```

It starts fetching all the stuff "at the same time", instead of waiting for each fetch to succeed (as it does today).

Reply:
(made the name before the implementation, and it made more sense back then... :sweat_smile:)

Reply:
Awesome!
Maybe call it something like: `hasDependantPrepares`? because this really says if there are any prepared compoments that depends on it.

Edit the comment to make a bit more sense and then we can merge this :100:.

A suggestion:
```jsx
/*
* Does this prepare function have to resolve before travelling further down the tree and
* executing prepare functions further down the tree?.
* If `<the name> == false`, the prepare function will be executed, but not awaited, allowing us
* to immedietly continue with the next prepare function.
* If `<the name> == true`, it will be executed AND awaited before continuing down the
* component tree.
* The unawaited prepare function is however awaited before prepare resolves, so that we have the
* result ready when rendering the html.
*/
```
_Or something like this._

Reply:
Hmm. Doesn't only apply to other prepares tho, can be eg. a random if inside a component that use the redux state..

Reply:
Also, it is important to note that this only applies to ssr, and not on the client side